### PR TITLE
Use fallback encodings when reading files

### PR DESCRIPTION
**Pull request by DanVanAtta**
Allows for ISO-8859 files to be read if files
are not encoded as ASCII (UTF-8). ISO-8859 characters
may not render correctly but the file will at least
be read (avoiding relatively obscure error scenarios).

Resolves: https://github.com/triplea-game/triplea/issues/9028

### CODE REVIEW

**Comment by Cernelius** [comment]
To be clear, I don't need this: all is working fine as of now. Maybe it's just better making the error more understandable (that the issue is the presence of unsupported characters and the file needs to be re-encoded as ASCII text).

It might be a bad idea to leave TripleA reading something having unnecessarily weird characters as it may be preferable keeping standards.

File: lib/java-extras/src/main/java/org/triplea/io/FileUtils.java [hunk 1]
```diff
@@ -180,13 +182,39 @@ public static URL toUrl(final Path file) {
...
+      } catch (final MalformedInputException e) {
+        log.info(
+            "Warning: "
+                + "unable to read file (character encoding problem), will next try ISO-8859_1: {}, {}",
+            fileToRead.toFile().getAbsolutePath(),
+            e.getMessage());
+      }
+
+      try {
+        return Optional.of(Files.readString(fileToRead, Charsets.ISO_8859_1));
```
Comment (RoiEXLab) [inline lib/java-extras/src/main/java/org/triplea/io/FileUtils.java:209]: I know that Latin-1 is still commonly used in waaay too many places. But I don't think this is the right fallback handling to perform.
If for some arbitrary reason the encoding uses Latin-9 which is very similar except for 8 characters this will probably still work without any issues but the characters will simply be the wrong ones.

Also `Charsets` is a class from the google libraries, the default java way to do it is `StandardCharsets`.

**Review by RoiEXLab — CHANGES_REQUESTED** [review]
I agree with @Cernelius here.
I don't think trying different charsets until no exception is thrown is the right move. Enforcing UTF-8 because it is the most widely used Unicode encoding out there is better in my opinion. Not only does it allow to use all characters of the Unicode standard, but this way we will never end up in a situation where a file is actually saved in a different encoding but happens to parse in a different encoding just with different characters.

**Comment by DanVanAtta** [comment]
@RoiEXLab how do you propose we enforce UTF-8? I don't think we can and we will struggle with map makers who are asking "WTF is a charset? How do I see it? What, I didn't change it? How do I make it UTF-8?". OTOH if we fall back to other charsets, we avoid that. Yeah, it would be better to require only UTF-8, but it introduces problems requiring that. Similarly, requiring only LF line endings in a file modified by a map maker would be good too, but again that introduces more problems than just allowing CRLF. Bottom line, while it might smell not to have UTF-8, the purity requirement adds unnecessary overhead and troubleshooting on ourt part, while not ideal, it can be made to just work.

File: lib/java-extras/src/main/java/org/triplea/io/FileUtils.java (hunk 1 above)
Comment (DanVanAtta) [reply lib/java-extras/src/main/java/org/triplea/io/FileUtils.java:209]: Good point on latin-8 vs latin-9.
- Better to have the content render than crash. 
- From what I've observed, any of the latin specific characters will not render well or at all correctly anyways. 
- I'd rather debug why a character is rendering weirdly than troubleshoot and train someone to use a UTF-8 text editor.

**Comment by DanVanAtta** [comment]
Re:
> Enforcing UTF-8 because it is the most widely used Unicode encoding out there is better in my opinion. 

If we make this a requirement, let me put it this way, then someone must be available to explain and train anyone that runs afoul of the UTF-8 requirement. Simply put, there is no such someone.

**Comment by RoiEXLab** [comment]
@DanVanAtta I think we are agreeing that UTF-8 is the desirable encoding for everything regarding this, right?

So I think we're actually discussing if we think that map makers are able to use UTF-8 by themselves.
I just checked: Every text editor on my machine (notepad.exe, Notepad++, Atom, VS Code, IntelliJ IDEA) uses UTF-8 by default. I actually wasn't sure if the default Windows editor did support proper UTF-8, but it looks like they changed it a while back so it is actually real UTF-8 without BOM.
That being said most if not all map makers shouldn't ever run into encoding issues if they are using popular & modern tooling.
Plus if they ever run into issues, regardless of type (crash, or different character displayed) we will end up with a ticket or a forum post regardless.
This change is only guaranteed to help in case you're using one of the additional 128 characters in ISO 8859-1 and does only prevent a crash, not necessarily fixing the issue using ascii extensions like ISO 8859 2 to 16, which are used even more rarely nowadays.

TL;DR I just don't think it's worth to maintain the extra code for an issue anyone will only ever have every couple years (I don't know if anyone before @Cernelius ever had this issue and reported it to us).

**Comment by DanVanAtta** [comment]
Yeah, we are debating whether to require map makers to use UTF-8 encodings or to be more permissive. 

There is no available support to help map makers debug UTF-8 issues. Assuming issues come up, the energy to do that support at the cost of 8'ish lines of code in a 'lib' module is not a good trade off. If we do require map makers to do UTF-8, then we need to document it, give some handy how-to guides, survey all major text editors, find out more about how this happened in the first place, and provide support whenever it comes up. This lib module is also new, and there are new use-cases for editting more file types than just XML, the fact an early adopter out of a very small number of early adopters (I"m not sure if there are any, so possible the *first* person), is a very bad sign. Even if the number is small, a rare problem is almost worse than a frequent, as it'll take even more time and round-trips to grok what is happening. I don't think this is up for debate, we have no full time staff, and even if we did, supporting *this* problem at the cost of characters rendering not at all, is not worth it.

**Comment by Cernelius** [comment]
I remember that, when I was working on WAW with @ron-murhammer, something like this came up that the game notes could not be read because of unsupported characters and was easily fixed by changing the unsupported characters.

All that is needed is that the message says that the file is not accepted because there are non-supported characters and preferably (but not necessarily) also telling exactly where (the line and such).

Side note, I reported this problem just because this "notes" file was a new thing, so I had no idea of what was working if anything. If it would have happened to anything else (which I know works), I would have easily guessed I must have written something wrong or unreadable without the need of any understandable warning.

In this case, however, I understand the unsupported character is some sort of unsupported white-space that looks like any other white-space to me (Right?), so I would have no clue. However, probably something very rare, as @RoiEXLab says.

And, yeah, chances are that I'm the first one trying to making a "notes" file manually.

I still don't really understand what a UTF-8 is, nor how I can tell my editor to be sure to generate a UTF-8 when I save the file, nor how I can see if an existent file is a UTF-8 or not, but I guess that as long as the file is UTF-8 and I don't add unsupported characters it should remain UTF-8.

The conclusion is that I'm really not seeing how it is a good idea to support substantially unacceptable things as that is burying problems which will accumulate overtime. Better to ensure that everything is correct by making sure that what is not does not work, so map-makers are forced to sort it out. Alternatively, always giving an error while making it work.

**Comment by Cernelius** [comment]
> I just checked: Every text editor on my machine (notepad.exe, Notepad++, Atom, VS Code, IntelliJ IDEA) uses UTF-8 by default. I actually wasn't sure if the default Windows editor did support proper UTF-8, but it looks like they changed it a while back so it is actually real UTF-8 without BOM.

Have you ever taken a look at my favourite text editor?
http://www.jedit.org/

I find it better than Notepad++ and anything else I tried.

**Comment by RoiEXLab** [comment]
@Cernelius I just installed it on my local windows machine and it looks like at least on windows the default charset is CP1252 (also called Windows-1252) which is very similar to ISO 8859-1 (Latin-1), but differs in some aspects. Judging by their "features" page where they advertise to support over 160 encoding and the fact that my java installation currently supports exactly 173 Encodings I would assume that jEdit simply relies on the encodings that come with java by default.
It probably also uses `Charset.defaultCharset()`, which - at least on my windows machine - returns said 'windows-1252' charset.
For you this means that in the settings menu you can simply change Cp1552 to UTF-8:
Screenshot

When using UTF-8 (or UTF-16 or UTF-32 for completeness) every Unicode character can be used, which includes 143,859 characters in Unicode 13 (standardized in 2020) which is probably every possible character you can imagine, including those of languages that are no longer spoken by anyone and of course emojis.
TL;DR: When using most non-unicode encodings you are limited to a set of roughly 255 characters, with Unicode when using UTF-8, UTF-16 or UTF-32 in all of their variants you can encode practically every character out there.

**Comment by RoiEXLab** [comment]
@Cernelius In case you're curious, I found a good comparison table that sums up the problem with character encodings pretty well:
https://en.wikipedia.org/wiki/Western_Latin_character_sets_(computing)#Comparison_table
The "Code Point" column represents the number the unicode standard assigns to this character. All the columns to the right represent the byte representation of said character from 0x00 / 0 to 0xFF / 255 in their respective charset.
There are obviously more than 255 characters worldwide, so some characters have to be skipped in some charsets in order to include all the relevant ones (which is very subjective, the reason why there are so many charsets, unicode tries to be the standard to rule them all).

Regardless I noticed I have been using the terms "charset" and "encoding" synonymously in this issue, which is not always the case.
Many charsets only come with a single encoding, like Latin-1. But the Unicode Standard is a charset that comes with many UTF-X encodings that arrange the unicode code points in different ways for storage each approach with pros and cons

**Comment by DanVanAtta** [comment]
Getting line numbers is not possible, a user will just be faced with the error message you observed. We might be able to say something more about needing to encode in UTF-8, I'm sure that will not be any more helpful though.

Using a different encoding is not wrong per-say, and we if we read the text content in the right encoding nothing is to say it won't render perfectly fine. The fault is arguably with the program for assuming UTF-8 when that is not always the case. UTF-8 is generally preferred, but it is not universally used. By the same token, latin encodings are relatively common, I wouldn't want us to bother going down the rabbit hole of supporting every encoding there is out there or spending even that much more time with it. Having a single fallback to support a very common encoding seems entirely reasonable.

If we do require only UTF-8, then we need a volunteer to improve the error messaging, update documentation, communicate to all users we will have this new requirement, and assist anyone in the future who has this problem. Nobody is able to make that commitment, it's a non-option. Even this dialog here has consumed more time than just allowing the encoding, I don't think there is a clear understanding of the negative impact (arguably none, just a few extra lines of code that someone in the future will have to read and because of where it is located likely never modify). So why exactly should we narrow the constraints the program for arguably no benefit just because UTF-8 "feels" clean? Second, why should we require only UTF-8 when it will cause support work for which there is no staff (and arguably it's a bad use of that staff's time if there were such a staff, even if payed, without evidence of significant negative impact, the application is being too narrowly constrained here).
